Modules/mathmodule.c

```diff
@@ -2217,41 +2217,63 @@ math_modf_impl(PyObject *module, double x)
    in that int is larger than PY_SSIZE_T_MAX. */
 
 static PyObject*
-loghelper(PyObject* arg, double (*func)(double))
+loghelper_int(PyObject* arg, double (*func)(double))
 {
     /* If it is int, do it ourselves. */
-    if (PyLong_Check(arg)) {
-        double x, result;
-        Py_ssize_t e;
+    double x, result;
+    Py_ssize_t e;
 
-        /* Negative or zero inputs give a ValueError. */
-        if (!_PyLong_IsPositive((PyLongObject *)arg)) {
-            PyErr_SetString(PyExc_ValueError,
-                            "math domain error");
-            return NULL;
-        }
+    /* Negative or zero inputs give a ValueError. */
+    if (!_PyLong_IsPositive((PyLongObject *)arg)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "math domain error");
+        return NULL;
+    }
 
-        x = PyLong_AsDouble(arg);
-        if (x == -1.0 && PyErr_Occurred()) {
-            if (!PyErr_ExceptionMatches(PyExc_OverflowError))
-                return NULL;
-            /* Here the conversion to double overflowed, but it's possible
-               to compute the log anyway.  Clear the exception and continue. */
-            PyErr_Clear();
-            x = _PyLong_Frexp((PyLongObject *)arg, &e);
-            if (x == -1.0 && PyErr_Occurred())
-                return NULL;
-            /* Value is ~= x * 2**e, so the log ~= log(x) + log(2) * e. */
-            result = func(x) + func(2.0) * e;
-        }
-        else
-            /* Successfully converted x to a double. */
-            result = func(x);
-        return PyFloat_FromDouble(result);
+    x = PyLong_AsDouble(arg);
+    if (x == -1.0 && PyErr_Occurred()) {
+        if (!PyErr_ExceptionMatches(PyExc_OverflowError))
+            return NULL;
+        /* Here the conversion to double overflowed, but it's possible
+           to compute the log anyway.  Clear the exception and continue. */
+        PyErr_Clear();
+        x = _PyLong_Frexp((PyLongObject *)arg, &e);
+        if (x == -1.0 && PyErr_Occurred())
+            return NULL;
+        /* Value is ~= x * 2**e, so the log ~= log(x) + log(2) * e. */
+        result = func(x) + func(2.0) * e;
     }
+    else
+        /* Successfully converted x to a double. */
+        result = func(x);
```

Comment:
I am not sure that it is worth to do. This code is not on common path.
```suggestion
    if (_PyLong_IsCompact((PyLongObject *)arg)) {
        x = (double)_PyLong_CompactValue((PyLongObject *)arg);
    }
    else {
        x = _PyLong_Frexp((PyLongObject *)arg, &e);
        if (x == -1.0 && PyErr_Occurred()) {
            return NULL;
        }
        if (e > DBL_MAX_EXP) {
            /* Here the conversion to double overflowed, but it's possible
               to compute the log anyway.
               Value is ~= x * 2**e, so the log ~= log(x) + log(2) * e. */
            result = func(x) + func(2.0) * e;
            return PyFloat_FromDouble(result);
        }
        else {
            x = ldexp(x, (int)e);
        }
    }
    /* Successfully converted x to a double. */
    result = func(x);
```

Reply:
One minor note: here again, PyErr_Occurred() - a useless expensive call.